### ISSUE TITLE
[HTML Error] HTML::Template::param() : attempt to set parameter 'groups' with a scalar - parameter is not a TMPL_VAR!

### ISSUE THREAD

**Issue by AIXMeister** [munin-monitoring/munin#610]
# Version

`munin 2.0.25`
# Prerequisites

On a fresh installation set the following parameters in `munin.conf`:

```
graph_strategy cgi
html_strategy cgi
```
# Incorrect behaviour

Browse the munin installation and load the same page multiple times. The following happens:
1. The page loads and displays normally
2. The page does not load. Instead this error is shown:

```
HTML::Template::param() : attempt to set parameter 'groups' with a scalar - parameter is not a TMPL_VAR! at /usr/share/perl5/vendor_perl/Munin/Master/HTMLOld.pm line 695.
```

This alternating behaviour - display normal, then display error - happens for any page on every reload of the page.
# Analysis

Line 55 in `munin/cgi/munin-cgi-html` loads the configuration:

```
$config = get_config(1);
```

The given parameter `1` tries to load the configuration from the cache.
With `html_strategy cgi` and `graph_strategy cgi` this cache does not exist/cannot be created.
# Workaround

Change the parameter to zero in line 55 in `munin/cgi/munin-cgi-html`:

```
$config = get_config(0);
```

Any page loads display normally if this workaround is in place.

**Comment by Nukesor** [comment]
This is actually still an open issue on 2.0.26 on arch linux.
Would love to see a bug fix in the next version!

HTML error:
```
Status: 500 Content-type: text/html
Software error:

HTML::Template::param() : attempt to set parameter 'groups' with a scalar - parameter is not a TMPL_VAR! at /usr/share/perl5/vendor_perl/Munin/Master/HTMLOld.pm line 695.

For help, please send mail to this site's webmaster, giving this error message and the time and date of the error.
```

**Comment by jorti** [comment]
I've this problem in Fedora with munin 2.0.33. The proposed workaround does not work for me.

**Comment by pbda** [comment]
I have the same problem with munin 2.0.33.1 on Raspbian  9 "stretch". It works every other time: once I get error message and once I get correct output. Change proposed by AIXMeister works for me if I restart fastcgi processes.

**Comment by sumpfralle** [comment]
I just did the following:
1. `apt install munin apache2 libapache2-mod-fcgid` on a Debian Stretch host
2. set `graph_strategy cgi` and `html_strategy cgi` in `/etc/munin/munin.conf`
3. configure apache according to [the example](http://guide.munin-monitoring.org/en/latest/example/webserver/apache-cgi.html)
4. repeatedly request the start page (or other pages)

I did not notice any errors.

Do you still encounter this issue? If yes: do the `*.storable` files exist on your system (e.g. `/var/lib/munin`)? Who is the owner and which permissions are set?

**Comment by pbda** [comment]
@sumpfralle 
I tried to revert back `munin/cgi/munin-cgi-html` to original file and it seems to work now !
I didn't use `apache` but `nginx`
```
# ls -l *.storable
-rw-r--r-- 1 munin munin    21085 mars   5 15:35 datafile.storable
-rw-r--r-- 1 munin www-data 81674 mars   5 15:28 htmlconf.storable
-rw-r--r-- 1 munin munin       11 nov.   9 22:40 limits.storable
-rw-r--r-- 1 munin munin     6997 mars   5 15:35 state-picon-picon.storable
-rw-r--r-- 1 munin munin     5552 mars   5 15:35 state-pitit-pitit.storable
```

**Comment by sumpfralle** [comment]
@pbda: thank you for taking a look again!

@AIXMeister, @Nukesor, @jorti: do you still experience this issue?
If yes: could you help me reproducing it?

**Comment by Nukesor** [comment]
@sumpfralle 
I just set up a new Server and encountered the problem again.
The proposed workaround in the original post works, and is only required for the first request.
Afterwards it can be changed to `get_config(1)` again and everything works fine.

I created a [gist](https://gist.github.com/Nukesor/1937b8553096b95ff45fe503809d1f7e) with my current nginx and munin/node config.
Munin is running locally for a single server.

**Comment by sumpfralle** [comment]
@Nukesor: thank you for the example configuration. Hopefully it helps me reproducing this behaviour.

Is there any chance that you could (without too much of a hassle) try it with a newer version of munin? 2.0.39 was just released a few minutes ago :)
(but I am not aware of any specific changes in this regard)

**Comment by trucheromayor** [comment]
Dear collegues,
I´ve the same issue. I´ve tried the get_config(1) solution... but it dosen´t work.
OS: Ubuntu 16.04
Munin version 2.0.25
Runing under apache2.

Server is running but web frontend is not working....:(

Any updates on this issue since the first post??

Thanks in advance 

Jon Garrido,

**Comment by sumpfralle** [comment]
@trucheromayor: just for making sure, that we are talking about the same issue ...
* Did you notice the same symptoms as the original submitter?
* Did you restart your webserver / your fastcgi daemon?
* Can you verify, whether it works for you in a newer version of munin? (we just reached 2.0.47)

**Comment by trucheromayor** [comment]
Thank you Lars for your response. 

About your indications:

- Not exactly having the same symptoms. In my case, the same response all the times. 
- Yes I did restart the webserver... 
- The version 2.0.25 is the latest in my stable ubuntu repository. As I can see it´s just the same version as  AIXMeister... .You´re right, may be we should add a backport repository, we will try to avoid this at this momment.

I´m doing a deeper research. I´ll feedback you...

Thanks again.

**Comment by artjomsimon** [comment]
Can confirm the issue, exactly as described by the original submitter, still exists in 2.0.47.
Fresh installation on Arch Linux led me here.

However, changing the suggested line in `munin/cgi/munin-cgi-html` to `$config = get_config(0);` does not resolve the issue for me.

**Comment by artjomsimon** [comment]
After reading through http://munin-monitoring.org/ticket/739 I checked my munin-node config.
It turns out I had configured a node that the master could not connect to.
Fixing that brought my munin instance to life.

**Comment by sumpfralle** [comment]
@artjomsimon: just for clarification - your problem was not related to the one described by the creator of this issue?

**Comment by kjetilho** [comment]
You get this error if the configuration contains *no* data.  It is not sufficient that a node is defined, there will also have to be collected data for at least one graph.

A strategic test for defined $config and an error message to somewhere would be nice.

**Comment by sumpfralle** [comment]
Thank you all for your thoughts and comments!

I fixed the issue in 48e1e0af6237ebb0863674163cf3835d9ee5a18e.
